beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go

```diff
@@ -197,86 +194,146 @@ func (vs *Server) BuildBlockParallel(ctx context.Context, sBlk interfaces.Signed
-	}).Debug("Broadcasting block")
-
-	if blk.Version() >= version.Deneb {
-		if !blinded {
-			dbBlockContents := req.GetDeneb()
-			if dbBlockContents == nil {
-				return nil, errors.New("signed beacon block contents is empty")
-			}
-			scs, err = buildBlobSidecars(blk, dbBlockContents.Blobs, dbBlockContents.KzgProofs)
-			if err != nil {
-				return nil, fmt.Errorf("could not build blob sidecars: %v", err)
-			}
-		}
-		for i, sc := range scs {
-			if err := vs.P2P.BroadcastBlob(ctx, uint64(i), sc); err != nil {
-				log.WithError(err).Error("Could not broadcast blob")
-			}
-			readOnlySc, err := blocks.NewROBlobWithRoot(sc, root)
-			if err != nil {
-				return nil, fmt.Errorf("could not create ROBlob: %v", err)
-			}
-			verifiedSc := blocks.NewVerifiedROBlob(readOnlySc)
-			if err := vs.BlobReceiver.ReceiveBlob(ctx, verifiedSc); err != nil {
-				log.WithError(err).Error("Could not receive blob")
-			}
-		}
+
+	return copiedBlock, sidecars, nil
+}
+
+// copyAndCastBlock creates a copy of a SignedBeaconBlock.
+func (vs *Server) copyAndCastBlock(block interfaces.SignedBeaconBlock) (interfaces.SignedBeaconBlock, error) {
+	copied, err := block.Copy()
+	if err != nil {
+		return nil, errors.Wrap(err, "copy failed")
+	}
+	castedBlock, ok := copied.(interfaces.SignedBeaconBlock)
+	if !ok {
+		return nil, errors.New("cast to SignedBeaconBlock failed")
 	}
+	return castedBlock, nil
+}
 
-	if err := vs.BlockReceiver.ReceiveBlock(ctx, blk, root); err != nil {
-		return nil, fmt.Errorf("could not process beacon block: %v", err)
+// handleUnblindedBlock processes unblinded beacon blocks.
+func (vs *Server) handleUnblindedBlock(block interfaces.SignedBeaconBlock, req *ethpb.GenericSignedBeaconBlock) ([]*ethpb.BlobSidecar, error) {
+	dbBlockContents := req.GetDeneb()
+	if dbBlockContents == nil {
+		return nil, nil
 	}
+	return buildBlobSidecars(block, dbBlockContents.Blobs, dbBlockContents.KzgProofs)
+}
 
-	log.WithField("slot", blk.Block().Slot()).Debugf(
-		"Block proposal received via RPC")
+// broadcastReceiveBlock broadcasts a block and handles its reception.
+func (vs *Server) broadcastReceiveBlock(ctx context.Context, block interfaces.SignedBeaconBlock, root [32]byte) error {
+	protoBlock, err := block.Proto()
+	if err != nil {
+		return errors.Wrap(err, "protobuf conversion failed")
+	}
+	if err := vs.P2P.Broadcast(ctx, protoBlock); err != nil {
+		return errors.Wrap(err, "broadcast failed")
+	}
 	vs.BlockNotifier.BlockFeed().Send(&feed.Event{
 		Type: blockfeed.ReceivedBlock,
-		Data: &blockfeed.ReceivedBlockData{SignedBlock: blk},
+		Data: &blockfeed.ReceivedBlockData{SignedBlock: block},
 	})
+	return vs.BlockReceiver.ReceiveBlock(ctx, block, root)
+}
 
-	return &ethpb.ProposeResponse{
-		BlockRoot: root[:],
-	}, nil
+// broadcastAndReceiveBlobs handles the broadcasting and reception of blob sidecars.
+func (vs *Server) broadcastAndReceiveBlobs(ctx context.Context, sidecars []*ethpb.BlobSidecar, root [32]byte) error {
```

Comment:
Why do we log when blob broadcast/receive fails, but return an error when the same happens for a block?

Reply:
Good catch. I'll return error instead